Merge OAVES flow

hqjenny opened the pull request.

angshuman-parashar commented:
A few requests:
- Clang is unhappy about the unused `total_min_traffic` variable in `mapper-thread.cpp`. Could you please remove the references in lines 709, 728, 825, 844. With this change I was able to make the code clear internal regressions.
- There's a lot of code replication in `MapperThread::PrintStats` and `MapperThread::PrintOAVESStats`. Could the two be merged?
- In general the above two methods break abstraction by boring into the guts of the problem shape and buffer storage levels. Is it possible to clean this up? E.g., try to put most of this code at a method call on `BufferLevel`. Then have the `Topology` call the methods for the levels it cares about (e.g., just the last level). And finally, `MapperThread` simply calls a clean method called `Print` something on Topology. Right now the abstraction breakage is quite ugly -- there's a comment called "assume the DRAM is the last level" on line 859, when `MapperThread` really shouldn't know anything about DRAM.
- If it's not too onerous, please also consider cleaning up unnecessary comments and conforming to the coding style that is predominantly used in the codebase.
- And lastly, please merge with the latest master (should be a very superficial merge).

Thanks!

hqjenny commented:
Hi Angshu, I addressed some of the issues, but still got question about whether we should/and how to store the `topology` for the index_factor_best mapping.

Changes:
- Removed `total_min_traffic` and `MapperThread::PrintStats`
- Added the `TotalUtilizedBytes` and `OperationalIntensity` functions to the buffer. 

Question: 
- The reason for not using the buffer and topology data structure is that the topology that the engine returns in [mapper_thread.cpp](https://github.com/NVlabs/timeloop/pull/191/commits/0d6e7e0460a14e995c07501a6fc7fe68e3cc4308#diff-bc0914b53a1e8ca65b7e17d209972b22934daaac74fb4a68e6ba03016faa254cL411) does not correspond to the   `index_factor_best` mapping. I only tried using buffer stats from the stats_.index_factor_best data structure. Shall we also buffer the topology in the `EvaluationResult`?

angshuman-parashar commented:
This looks great now, thanks very much!
